File: android_mobile/src/main/java/com/alexstyl/specialdates/support/AskForSupport.java
package com.alexstyl.specialdates.support;

import android.content.Context;
import android.content.Intent;
import android.text.format.DateUtils;
import android.util.Log;

public class AskForSupport {

    private static final long RETRY_INTERVAL = DateUtils.DAY_IN_MILLIS * 1;
    private CallForRatingPreferences preferences;

    public AskForSupport(Context context) {
        this.preferences = new CallForRatingPreferences(context);
    }

    public boolean shouldAskForRating() {
        return !preferences.hasUserRated() && isTimeToAskAgain();
    }

    private boolean isTimeToAskAgain() {
        if (preferences.isTriggered()) {
            preferences.resetTrigger();
            return true;
        }
        long timeSinceLastRate = System.currentTimeMillis() - preferences.lastAskTimeAsked();
        return timeSinceLastRate > RETRY_INTERVAL;
    }

    public void onRateEnd() {
        preferences.setHasUserRated(true);
    }

    public void requestForRatingSooner() {
        preferences.triggerNextTime();
    }

    public void askForRatingFromUser(Context context) {
        context.startActivity(new Intent(context, RateDialog.class));
        preferences.setLastAskTimedNow();
    }
}

File: android_mobile/src/main/java/com/alexstyl/specialdates/support/AskForSupport.kt
package com.alexstyl.specialdates.support

import android.app.Activity
import android.content.Intent
import android.text.format.DateUtils

class AskForSupport(private val preferences: CallForRatingPreferences) {

    private val isTimeToAskAgain: Boolean
        get() {
            if (preferences.isTriggered) {
                preferences.resetTrigger()
                return true
            }
            val timeSinceLastRate = System.currentTimeMillis() - preferences.lastAskTimeAsked()
            return timeSinceLastRate > RETRY_INTERVAL
        }

    fun shouldAskForRating(): Boolean {
        return !preferences.hasUserRated() && isTimeToAskAgain
    }

    fun onRateEnd() {
        preferences.setHasUserRated(true)
    }

    fun requestForRatingSooner() {
        preferences.triggerNextTime()
    }

    fun askForRatingFromUser(activity: Activity) {
        val intent = Intent(activity, RateDialog::class.java)
        activity.startActivity(intent)
        preferences.setLastAskTimedNow()
    }

    companion object {
        private const val RETRY_INTERVAL = DateUtils.DAY_IN_MILLIS
    }
}
